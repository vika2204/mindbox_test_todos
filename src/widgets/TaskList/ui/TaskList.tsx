import {Button, Checkbox, CheckboxChangeEvent, Divider, Flex, Table, TableColumnsType} from 'antd';
import React, {useEffect, useState} from "react";
import {Task, TASK_STATUS} from "../../../entities/Task/model";
import {TaskService} from "../../../entities/Task/api";
import {Filters} from "./Filters.tsx";
import AddTask from "./AddTask.tsx";

export function TaskList() {
    const [filterByStatus, setFilterByStatus] = useState<TASK_STATUS|null>(null);
    let [tasks, setTasks] = useState<Task[]>(TaskService.loadTasks());

    const columns: TableColumnsType<Task> = [
        {
            title: 'Task Name',
            dataIndex: 'taskName',
            render: (text, record) => (
                <>
                    <Checkbox
                        onChange={(e) => handleCheckboxChange(e, record.key)}
                        checked={record.status === TASK_STATUS.DONE}
                    >{record.status === TASK_STATUS.DONE ? <s>{text}</s> : text}</Checkbox>
                </>
            ),
        },
    ];

    function handleCheckboxChange(e: CheckboxChangeEvent, taskKey: React.Key) {
        setTasks(
            tasks.map((task: Task) => {
                if (task.key !== taskKey) {
                    return task;
                }

                task.status = e.target.checked ? TASK_STATUS.DONE : TASK_STATUS.TODO;
                return task;
            })
        )
    }

    function createTask(taskName: string): void {
        let newTask: Task = {
            key: tasks.length + 1,
            taskName: taskName,
            status: TASK_STATUS.TODO,
        }

        setTasks([newTask, ...tasks]);
    }

    function clearCompletedTasks(): void {
        setTasks(
            tasks.filter((task: Task) => task.status !== TASK_STATUS.DONE)
        )
    }

    useEffect((): void => {
        TaskService.saveTasks(tasks);
    }, [tasks]);

    return (
        <>
            <AddTask onAdd={(tt) => createTask(tt)}/>
            <Divider></Divider>
            <Filters currentFilter={filterByStatus} setFilter={setFilterByStatus}/>
            <Divider></Divider>
            <Table<Task>
                columns={columns}
                dataSource={[...tasks.filter((task: Task) => filterByStatus ? task.status === filterByStatus : true)]}
                showHeader={false}
                pagination={{
                    hideOnSinglePage: true,
                    pageSize: 10
                }}
                footer={() => <Flex justify={"space-between"}>{
                    tasks.filter((task: Task) => task.status === TASK_STATUS.TODO).length
                } left <Button onClick={clearCompletedTasks}>Clear completed</Button></Flex>}
            />
        </>
    );
}

