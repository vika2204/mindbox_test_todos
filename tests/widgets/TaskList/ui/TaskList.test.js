import renderer from 'react-test-renderer';
import {TaskList} from "widgets/TaskList/ui/TaskList.tsx";
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

it('все компоненты на месте', () => {
   Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
         matches: false,
         media: query,
         onchange: null,
         addListener: jest.fn(), // Deprecated
         removeListener: jest.fn(), // Deprecated
         addEventListener: jest.fn(),
         removeEventListener: jest.fn(),
         dispatchEvent: jest.fn(),
      })),
   });

   const component = renderer.create(
       <TaskList/>
   );
   let tree = component.toJSON();
   expect(tree).toMatchSnapshot();
});

it('задачи добавляются', () => {
   render(<TaskList />);

   const inputElement = screen.getByPlaceholderText('What needs to be done?');
   const addButton = screen.getByText('Add');

   fireEvent.change(inputElement, { target: { value: 'Новая задача' } });
   fireEvent.click(addButton);

   const taskElement = screen.getByText('Новая задача');
   expect(taskElement).toBeInTheDocument();
});

it('фильтры применяются', () => {
   render(<TaskList />);

   const inputElement = screen.getByPlaceholderText('What needs to be done?');
   const addButton = screen.getByText('Add');
   const completedBtn = screen.getByText('Completed');
   const activeBtn = screen.getByText('Active');

   fireEvent.change(inputElement, { target: { value: 'Сделать тестовое задание' } });
   fireEvent.click(addButton);

   let taskElement = screen.getByText('Сделать тестовое задание');
   expect(taskElement).toBeInTheDocument();

   fireEvent.click(completedBtn);
   expect(() => screen.getByText('Сделать тестовое задание')).toThrow();

   fireEvent.click(activeBtn);
   taskElement = screen.getByText('Сделать тестовое задание');
   expect(taskElement).toBeInTheDocument();
});

it('статус задачи меняется при нажатии на галочку', () => {
   render(<TaskList />);
   const inputElement = screen.getByPlaceholderText('What needs to be done?');
   const addButton = screen.getByText('Add');
   const activeBtn = screen.getByText('Active');
   const completedBtn = screen.getByText('Completed');

   fireEvent.click(activeBtn);

   fireEvent.change(inputElement, { target: { value: 'Записаться на бадминтон' } });
   fireEvent.click(addButton);

   let taskElement = screen.getByText('Записаться на бадминтон');
   expect(taskElement).toBeInTheDocument();
   fireEvent.click(taskElement);

   expect(() => screen.getByText('Записаться на бадминтон')).toThrow();

   fireEvent.click(completedBtn);
   taskElement = screen.getByText('Записаться на бадминтон');
   expect(taskElement).toBeInTheDocument();
});


