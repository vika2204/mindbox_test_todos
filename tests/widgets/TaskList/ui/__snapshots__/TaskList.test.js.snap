// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`все компоненты на месте 1`] = `
[
  <div
    className="ant-space-compact css-dev-only-do-not-override-1htoz2s"
    style={
      {
        "width": "100%",
      }
    }
  >
    <input
      autoFocus={true}
      className="ant-input ant-input-lg css-dev-only-do-not-override-1htoz2s ant-input-outlined ant-input-compact-item"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onCompositionEnd={[Function]}
      onCompositionStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder="What needs to be done?"
      style={{}}
      type="text"
      value=""
    />
    <button
      className="ant-btn css-dev-only-do-not-override-1htoz2s ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-lg ant-btn-compact-item ant-btn-compact-last-item"
      disabled={false}
      onClick={[Function]}
      style={{}}
      type="button"
    >
      <span>
        Add
      </span>
    </button>
  </div>,
  <div
    className="ant-divider css-dev-only-do-not-override-1htoz2s ant-divider-horizontal"
    role="separator"
    style={{}}
  />,
  <div
    className="ant-space css-dev-only-do-not-override-1htoz2s ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small"
    style={
      {
        "width": "100%",
      }
    }
  >
    <div
      className="ant-space-item"
    >
      <span
        className="ant-tag ant-tag-checkable ant-tag-checkable-checked css-dev-only-do-not-override-1htoz2s"
        onClick={[Function]}
        style={{}}
      >
        All
      </span>
    </div>
    <div
      className="ant-space-item"
    >
      <span
        className="ant-tag ant-tag-checkable css-dev-only-do-not-override-1htoz2s"
        onClick={[Function]}
        style={{}}
      >
        Active
      </span>
    </div>
    <div
      className="ant-space-item"
    >
      <span
        className="ant-tag ant-tag-checkable css-dev-only-do-not-override-1htoz2s"
        onClick={[Function]}
        style={{}}
      >
        Completed
      </span>
    </div>
  </div>,
  <div
    className="ant-divider css-dev-only-do-not-override-1htoz2s ant-divider-horizontal"
    role="separator"
    style={{}}
  />,
  <div
    className="ant-table-wrapper css-dev-only-do-not-override-1htoz2s"
    style={{}}
  >
    <div
      className="ant-spin-nested-loading css-dev-only-do-not-override-1htoz2s"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-empty css-dev-only-do-not-override-1htoz2s"
        >
          <div
            className="ant-table-container"
          >
            <div
              className="ant-table-content"
              onScroll={[Function]}
              style={{}}
            >
              <table
                style={
                  {
                    "tableLayout": "auto",
                  }
                }
              >
                <colgroup />
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-placeholder"
                    style={
                      {
                        "display": null,
                      }
                    }
                  >
                    <td
                      className="ant-table-cell"
                      colSpan={null}
                      onMouseEnter={[Function]}
                      onMouseLeave={[Function]}
                      rowSpan={null}
                      style={{}}
                    >
                      <div
                        className="css-dev-only-do-not-override-1htoz2s ant-empty ant-empty-normal"
                        style={{}}
                      >
                        <div
                          className="ant-empty-image"
                          style={{}}
                        >
                          <svg
                            height="41"
                            viewBox="0 0 64 41"
                            width="64"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <title>
                              No data
                            </title>
                            <g
                              fill="none"
                              fillRule="evenodd"
                              transform="translate(0 1)"
                            >
                              <ellipse
                                cx="32"
                                cy="33"
                                fill="#f5f5f5"
                                rx="32"
                                ry="7"
                              />
                              <g
                                fillRule="nonzero"
                                stroke="#d9d9d9"
                              >
                                <path
                                  d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                />
                                <path
                                  d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  fill="#fafafa"
                                />
                              </g>
                            </g>
                          </svg>
                        </div>
                        <div
                          className="ant-empty-description"
                          style={{}}
                        >
                          No data
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
          <div
            className="ant-table-footer"
          >
            <div
              className="ant-flex css-dev-only-do-not-override-1htoz2s ant-flex-justify-space-between"
              style={{}}
            >
              0
               left 
              <button
                className="ant-btn css-dev-only-do-not-override-1htoz2s ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                disabled={false}
                onClick={[Function]}
                style={{}}
                type="button"
              >
                <span>
                  Clear completed
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
